{"cells":[{"cell_type":"markdown","metadata":{},"source":["Group Name: AG 03.\n","\n","Student Name (Student ID):\n","\n","1. xxxx xxxxx (xxxxxxx)\n","\n","2. xxxx xxxxx (xxxxxxx)\n","\n","3. xxxx xxxxx (xxxxxxx)"]},{"cell_type":"markdown","metadata":{"id":"yZOz6Y_cLGOT"},"source":["# Question 2\n","\n","## Introduction to question 2\n","\n","In the second question of this assignment, we will explore the use of local search in genome assembly.\n","\n","We will use local search to assemble (construct) a large part of the nucleotide sequence of the monkeypox virus, which has been downloaded from the National Center for Biotechnology Information in the United States. Please note that no additional or specialized knowledge of biology or bioinformatics is required for this assignment. (Actually, the technical specifics of bioinformatics have been adapted and simplified for the purposes of this computer science assignment, so if you are a biologist, please do not apply preexisting knowledge to solve the problem. Furthermore, you should not attempt to search up the genome on genomic databases to \"guess\" the actual sequence, since we are more interested in your coding methodology rather than your attempts at reproducing a known sequence.)\n","\n","This is an introductory computer science assignment and not a bioinformatics assignment; we are simply using bioinformatics as a use case to illustrate the applicability of local search to the natural sciences. Therefore, no knowledge of bioinformatics is assumed or required. In the paragraphs that follow, I will give a short crash course which will cover all the domain knowledge you will need to know in order to tackle this problem.  \n","\n","For technical reasons, when we analyze the nucleotide sequence (genome) of a virus, we usually cannot “read” it in one fell swoop. We have to read the genome in parts, because the genome is usually too long for the machine to read in a single sitting. To simplify things, a “read” is a single view of part of the genome; think of it as a SUBSTRING, a partial view of the whole genome. After we have generated multiple reads of a genome, we then have to “stitch”, or combine, the different reads of the genome together. This process of stitching up reads of a genome into the final sequence is known as genome assembly. However, the different reads of the genome cannot just be concatenated like usual string concatenation. It’s not a situation where you have one read, “Hello”, and another read, “World”, and all you need to do is concatenate both strings together to make “Hello World”. Among other reasons, there are two major reasons why you can’t do so:\n","\n","1. You do not know which read came first. The reads are not ordered. How do you know “Hello” came after “World”? The answer is that you don’t. Imagine how complicated this situation might be if you had more than two reads. (This is indeed our situation, where we have $n$ reads, and $n>>2$.)\n","\n","2. One read may contain a substring contained in another read. Specifically, without loss of generality, part of the ending $x$ characters of a read (i.e., suffix) might also be found in the starting $x$ positions (i.e., prefix) of another read.\n","\n","- A computer scientist usually creates opportunities from problems. While this may be a “problem” in that you just can’t concatenate two strings blindly, the fact that strings contain shared “substrings” is actually a very helpful clue that you can use to “join” strings together. \n","\n","- Note that the choice of the value of $x$ could be a hyperparameter decided by the computer scientist.\n","\n","## Your tasks\n","\n","In this part of the assignment, you will work with (simulated) reads that I have generated from the nucleotide sequence of the monkeypox virus. In reality, bioinformatics is far more complicated, but here we will work with a simplified situation. Your task is to examine the reads that I have provided for you, and from there “infer” the nucleotide sequence that might have produced those reads. \n","\n","The reads are provided in the csv file `data.csv` which simply provides a list of unique strings. Note that you should NOT assume any particular ordering of the strings in this dataframe. In fact, the strings have already been shuffled randomly. \n","\n","NOTE: You are not allowed to use `pandas` or any other libraries apart from the Python STL to load the csv file.\n","\n","### Task A (3 marks): \n","\n","Create a directed graph. The nodes in the graph are the strings in the list of reads. An edge should be drawn FROM read A TO read B if and only if a suffix (of length $x$) of read A is also a prefix (obviously, also of length $x$) of read B. For the purposes of the assignment, limit the value of $x$ to between 5 and 30, both inclusive. That is, to be clear, $5\\leq x\\leq 30$. The weight of an edge between read A and read B should be the NEGATED value of $x$, i.e. $-x$. \n","\n","In your Jupyter notebook, please report the number of edges in your graph. Provide a barplot or histogram which shows the number of edges with different weights or weight categories. In this task, you are free to use plotting libraries such as `matplotlib` or `seaborn` to plot this graph.\n","\n","As an example, if read A is \"TACTAGT\" and read B is \"TAGTCCCCT\", then an edge is drawn FROM read A TO read B (i.e., $A \\rightarrow B$) with weight of $-4$. This is because the 4-suffix \"TAGT\" is also the 4-prefix of read B; in other words, the last 4 characters of read A (a substring of length 4) overlap with the first 4 characters of read B (a substring of length 4).\n","\n","### Task B (7 marks): \n","\n","From Task A, you now have a graph which shows connections between reads based on how they overlap, in theory you could draw a path through the graph and thereby derive the full sequence (genome).\n","\n","Task B asks you to use local search method(s) to determine a path through this directed graph of strings. \n","\n","- You are expected to use simulated annealing and tune the relevant configuration settings and hyperparameters. The minimum requirement is to implement simulated annealing.\n","\n","- Explain tha rationale behind the choice of scheduling strategy and parameters.\n","\n","- However, you may also explore other search methods in addition to simulated annealing. Marks will be awarded for effort.\n","\n","Note the following constraints:\n","\n","1. The path has to go through each and every vertex exactly once. For computer scientists, this constraint is reminiscent of the \"Traveling Salesman's Problem\", except that unlike TSP, we should not need to go back to the starting vertex again. \n","\n","2. For the purposes of neighbor generation / action selection at each node, bear in mind that a path through the graph which minimizes the total number of nucleotides in the assembled sequence is the preferred path. To state that another way, the assembled sequence should be derived from a path that goes through EACH and EVERY vertex exactly once, however we want this assembled sequence to be AS SHORT AS POSSIBLE.\n","\n","3. You are not given the starting (source/origin) or ending (destination) vertex.\n","\n","4. For avoidance of ambiguity, no cycles are allowed. You must not visit a vertex more than once.\n","\n","5. You are not allowed to use any libraries apart from the Python Standard Library.\n","No import statements which import libraries outside of the Python STL should be found within your answer for Task B.\n","\n","Please remember to report the assembled sequence that you obtain. Although it would be great if you can come up with a good sequence, please feel reassured that we are more interested in your APPROACH to the problem, and so you can potentially get a reasonable score on this task even if your solution is \"wrong\". It is the process, rather than the result, which matters more."]},{"cell_type":"code","execution_count":72,"metadata":{"id":"oBnbZpd2bt1N"},"outputs":[],"source":["# Problem Class\n","class Problem:\n","    \"\"\"The abstract class for a formal problem. A new domain subclasses this,\n","    overriding `actions` and `results`, and perhaps other methods.\n","    The default heuristic is 0 and the default action cost is 1 for all states.\n","    When you create an instance of a subclass, specify `initial`, and `goal` states \n","    (or give an `is_goal` method) and perhaps other keyword args for the subclass.\"\"\"\n","\n","    def __init__(self, initial=None, goal=None, **kwds): \n","        self.__dict__.update(initial=initial, goal=goal, **kwds) \n","        \n","    def actions(self, state):        raise NotImplementedError\n","    def result(self, state, action): raise NotImplementedError\n","    def is_goal(self, state):        return state == self.goal\n","    def action_cost(self, s, a, s1): return 1\n","    def h(self, node):               return 0\n","    \n","    def __str__(self):\n","        return '{}({!r}, {!r})'.format(\n","            type(self).__name__, self.initial, self.goal)"]},{"cell_type":"code","execution_count":84,"metadata":{},"outputs":[],"source":["# Use the following Node class to generate search tree\n","import math\n","class Node:\n","    \"A Node in a search tree.\"\n","    def __init__(self, state, parent=None, action=None, path_cost=0):\n","        self.__dict__.update(state=state, parent=parent, action=action, path_cost=path_cost)\n","\n","    def __repr__(self): return '<{}>'.format(self.state)\n","    def __len__(self): return 0 if self.parent is None else (1 + len(self.parent))\n","    def __lt__(self, other): return self.path_cost < other.path_cost \n","\n","    #Extensions to Node Class\n","    def expand(problem, node):\n","        \"Expand a node, generating the children nodes.\"\n","        s = node.state\n","        for action in problem.actions(s):\n","            s1 = problem.result(s, action)\n","            cost = node.path_cost + problem.action_cost(s, action, s1)\n","            return Node(s1, node, action, cost)\n","            \n","\n","    def path_actions(node):\n","        \"The sequence of actions to get to this node.\"\n","        if node.parent is None:\n","            return []  \n","        return path_actions(node.parent) + [node.action]\n","\n","\n","    def path_states(node):\n","        \"The sequence of states to get to this node.\"\n","        if node in (cutoff, failure, None): \n","            return []\n","        return path_states(node.parent) + [node.state]"]},{"cell_type":"code","execution_count":74,"metadata":{"id":"B2-y3Nt9ccxj"},"outputs":[],"source":["# Code to generate neighbours, value of states, etc.\n","class TSP(Problem):\n","    #Implement TSP class here\n","    pass"]},{"cell_type":"markdown","metadata":{},"source":["## Part 1 - Implementation of a graph structure of data.csv. \n","\n","\n","Input Data: to be structured into a dictionary using {index:read_sequence} \n","\n","### Graph Datatype\n","Directed Graph: to be built as an adjacency matrix\n","\n","*Example of Directed Graph Data Format*\n","```\n","Adjacency Dictionary \n","directed_graph = {\n","                    A : [(B,4), (C,1), ..., (F,3)], \n","                    B : [(E,5), (H,3), ..., (J,3)], \n","                    ...\n","                    }  #letters to be index and numbers to be overlap. \n","```\n","\n","```\n","Adjacency List \n","directed_graph = [\n","                    [A, B, 5], \n","                    [A, C, 3], \n","                    [D, E, 2],\n","                    ....\n","                    ]  #letters to be index. \n","```\n","\n","\n","For simplicity, decision to build graph as an adjacency list first. To build the graph, we will iterate through edge weights and count the numbers. This to be done as a once off, static output. For future optimization, this could be implemented as a lookup table of suffixes and prefixes in the format {suffix/prefix: id1, id2, idx} and set overlaps used to generate dynamic outputs. "]},{"cell_type":"code","execution_count":75,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["[[0,\n","  'CTTGAATTGGTTCCTGGTATCATTAGGATCTCTGTCTCTCAACATCTGTTTAAGTTCATCGAGAACCACCTCCTCATTTTCCAGATAGTCAAACATTTTGACTGAATAGAAGTGAATGAGCTACTGTGAACTCTATACACCCGCACAACTAATGTCATTAAATATCATTTTTGAATGTATTTATACCATGTCAAAAACTTGTACAATTATTAATAAAAATAATTAGTGTTTAAATTTTACCAGTTCCAGATTTTACACCTCCGTTAACACCTCCATTAACCCCACTTTTTACACCACTGGACGATCCTCCTCCCCACATTCCACTGCCACTAGATGTATAAGTTTTAGATCCTTTATTACTACCATCATGTCCATGGATAAAGACACTCCACATGCCGCCACTACTACCCCCT'],\n"," [1,\n","  'ATCTTTAACGAACATATACCTAGATGGTTATTTACTAACAGACATTTTTTCAAGATCTATTGACAATAACTCCTATAGTTTCCACATCAACCAAGTAATGATCATCTATTGTTATATAACAATAACATAACTCTTTTCCATTTTTATCAGTATCTATATCAACGTCGTTGTAGTGAATAGTAGTCATTGATCTATTATATGAAACGGATATGTCTAGTTAATATTTTCTTTGATTTAAAGTCTATAGTCTTTACAAACATAATATCCTTATCCGACTTTATATTTCCTGTAGGGTGGCATAATTTTATTCTGCCTCCACAATCAGTGTTTCCAAATATATTACTAGACAATATTCCATATAGT'],\n"," [2,\n","  'TTGTACATGTAATGATTTAAAATGTGTAGTCATGCTTATTGATAAAGATCTAAAAATTAAAGCGGGTCCTCGGTACGTGCTTAACGCTATTAGTCCTCATGCCTATGATGTTTTTAGAAAATCTAATAACTTGAAAGAGATAATAGAAAATGCAGCTAAACAAAATCTAGACTCTATATCTATTTCTGTTATGACTCCAATTAATCCCATGTTAGCGGAATCATGTGATTCTGTCAATAAGGCGTTTAAAAAATTTCCATCAGGAATGTTTGCGGAAGTCAAATACGATGGTGAAAGAGTACAAGTTCATAAAAAAAATAACGA'],\n"," [3,\n","  'TACTAGATTTTATCTCTAGCGAGATTGTTTAGAATCATTTATCATAACTATGTTTAATAAATTCATCAACGAATATCGATAAAGACCTCTTGTAATTCGAGTATAGGAAGTAGTATTACCATATCAACTTCCGAGTTAACAATTACTCTAAAACATGAGGATTGTACTCCTGTCTTTATTGGAGATCACTATTTAGTCGTTGATAAACTAGTAACCTCAGGTTTCTTTACAAACGATAAAGTACAACATCAAGACCTCACAACACAGTGCAAGATTAATCTAGAAATCAAATGTAATTCTGGAGGAGA'],\n"," [4,\n","  'CTGTCAGAAAATACCATCCAAGCCAATATCTTTACTTTTGTTCATCAGACGTTCCGGAAAGAGGTCCTCAGGTAGGTTTAGTATCTCAATTGTCTGTCTTGAGTTCCATTACAAATATACTAACGTCTGAGTATTTGGATTTGGAAAAGAAAATTTGTGAGTATATCAGATCATATTATAAAGATGATATAAGTTACTTTGAAACAGGATTTCCAATCACTATAGAAAATGCTCTAGTCGCATCTCTTAATCCAAATATGATATGTGATTTTGTAACTGACTTTAGACGTAGAAAACGGATGGGATTCTTCGGTAACT']]\n"]}],"source":["import csv\n","from pprint import pprint\n","\n","# Import the data, discarding column 2 because its similar to column 1\n","data_reads = [] \n","with open ('data.csv', 'r') as csv_file:\n","    for line in csv.reader(csv_file, delimiter=','):\n","        data_reads.append([line[0],line[2]])\n","#print (data)\n","\n","data_reads = [[int(row[0]), row[1]] for row in data_reads[1:]] #discard the header row, convert types\n","pprint (data_reads[:5]) #🔴 since starts from zero, this can be a 1dimensional list accessed by index\n"]},{"cell_type":"code","execution_count":94,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["[[0, 302, -5],\n"," [0, 224, -16],\n"," [1, 511, -5],\n"," [1, 344, -6],\n"," [1, 228, -18],\n"," [2, 223, -22],\n"," [3, 548, -5],\n"," [3, 355, -24],\n"," [4, 436, -17],\n"," [5, 102, -5]]\n"]}],"source":["# Find the overlapping prefix and suffixes \n","def overlap(readline: list, readlist: list, graph_output: list = None) -> str: # takes\n","    '''\n","    Lists should be in format [id, string]\n","    Takes a readline suffix and compares it against a readlist prefix for overlap in range 5 to 30\n","    output: \n","    - matches are it appends it to the graph_ouput list \n","    - format: [readline ID, readlist item ID, weight] i.e. A -> B, Weight \n","    '''\n","    for i in range (5, 31): #🔴 (1) double check if range and slices are correct. (2), this can be made more efficient: do i in reverse and add an idx_matched as wel \n","        \n","        for line in readlist: \n","            suffix = readline[1][-i:]\n","            prefix = line[1][:i]\n","            #print (f'{suffix}, {prefix}')\n","            if suffix == prefix:\n","                #print (f'match! {prefix} : {line[0]}-{suffix}')\n","                graph_output.append([readline[0], line[0], i * -1])\n","                #graph_adj_list.append([print ('match!')\n","\n","graph_adj_list = [] \n","for i in range(len(data_reads)):\n","    overlap(data_reads[i], data_reads, graph_adj_list)\n","\n","pprint (graph_adj_list[:10])"]},{"cell_type":"code","execution_count":95,"metadata":{},"outputs":[{"data":{"application/vnd.plotly.v1+json":{"config":{"plotlyServerURL":"https://plot.ly"},"data":[{"alignmentgroup":"True","hovertemplate":"weights=%{x}<br>count=%{y}<extra></extra>","legendgroup":"","marker":{"color":"#636efa","pattern":{"shape":""}},"name":"","offsetgroup":"","orientation":"v","showlegend":false,"textposition":"auto","type":"bar","x":[-5,-16,-6,-18,-22,-24,-17,-15,-7,-12,-25,-23,-20,-11,-9,-27,-30,-13,-29,-21,-14,-10,-26,-19,-28,-8],"xaxis":"x","y":[633,25,200,36,23,30,27,27,59,33,37,34,39,33,12,23,30,27,31,26,22,27,29,22,23,17],"yaxis":"y"}],"layout":{"barmode":"relative","legend":{"tracegroupgap":0},"margin":{"t":60},"template":{"data":{"bar":[{"error_x":{"color":"#2a3f5f"},"error_y":{"color":"#2a3f5f"},"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"bar"}],"barpolar":[{"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"barpolar"}],"carpet":[{"aaxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"baxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"type":"carpet"}],"choropleth":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"choropleth"}],"contour":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1,"#f0f921"]],"type":"contour"}],"contourcarpet":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"contourcarpet"}],"heatmap":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1,"#f0f921"]],"type":"heatmap"}],"heatmapgl":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1,"#f0f921"]],"type":"heatmapgl"}],"histogram":[{"marker":{"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"histogram"}],"histogram2d":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1,"#f0f921"]],"type":"histogram2d"}],"histogram2dcontour":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1,"#f0f921"]],"type":"histogram2dcontour"}],"mesh3d":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"mesh3d"}],"parcoords":[{"line":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"parcoords"}],"pie":[{"automargin":true,"type":"pie"}],"scatter":[{"fillpattern":{"fillmode":"overlay","size":10,"solidity":0.2},"type":"scatter"}],"scatter3d":[{"line":{"colorbar":{"outlinewidth":0,"ticks":""}},"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatter3d"}],"scattercarpet":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattercarpet"}],"scattergeo":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattergeo"}],"scattergl":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattergl"}],"scattermapbox":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattermapbox"}],"scatterpolar":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterpolar"}],"scatterpolargl":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterpolargl"}],"scatterternary":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterternary"}],"surface":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1,"#f0f921"]],"type":"surface"}],"table":[{"cells":{"fill":{"color":"#EBF0F8"},"line":{"color":"white"}},"header":{"fill":{"color":"#C8D4E3"},"line":{"color":"white"}},"type":"table"}]},"layout":{"annotationdefaults":{"arrowcolor":"#2a3f5f","arrowhead":0,"arrowwidth":1},"autotypenumbers":"strict","coloraxis":{"colorbar":{"outlinewidth":0,"ticks":""}},"colorscale":{"diverging":[[0,"#8e0152"],[0.1,"#c51b7d"],[0.2,"#de77ae"],[0.3,"#f1b6da"],[0.4,"#fde0ef"],[0.5,"#f7f7f7"],[0.6,"#e6f5d0"],[0.7,"#b8e186"],[0.8,"#7fbc41"],[0.9,"#4d9221"],[1,"#276419"]],"sequential":[[0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1,"#f0f921"]],"sequentialminus":[[0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1,"#f0f921"]]},"colorway":["#636efa","#EF553B","#00cc96","#ab63fa","#FFA15A","#19d3f3","#FF6692","#B6E880","#FF97FF","#FECB52"],"font":{"color":"#2a3f5f"},"geo":{"bgcolor":"white","lakecolor":"white","landcolor":"#E5ECF6","showlakes":true,"showland":true,"subunitcolor":"white"},"hoverlabel":{"align":"left"},"hovermode":"closest","mapbox":{"style":"light"},"paper_bgcolor":"white","plot_bgcolor":"#E5ECF6","polar":{"angularaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"bgcolor":"#E5ECF6","radialaxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"scene":{"xaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"},"yaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"},"zaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"}},"shapedefaults":{"line":{"color":"#2a3f5f"}},"ternary":{"aaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"baxis":{"gridcolor":"white","linecolor":"white","ticks":""},"bgcolor":"#E5ECF6","caxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"title":{"x":0.05},"xaxis":{"automargin":true,"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","zerolinewidth":2},"yaxis":{"automargin":true,"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","zerolinewidth":2}}},"xaxis":{"anchor":"y","domain":[0,1],"title":{"text":"weights"}},"yaxis":{"anchor":"x","domain":[0,1],"title":{"text":"count"}}}}},"metadata":{},"output_type":"display_data"}],"source":["# Charting\n","import plotly.express as px #using plotly because its prettier, easier and more extensible than seaborn. :D \n","#🔴 Change to seaborn \n","from collections import Counter\n","\n","histogram = dict(Counter([line[2] for line in graph_adj_list]))\n","histogram = {'weights': list(histogram.keys()), 'count': list(histogram.values()) } #ref: https://stackoverflow.com/questions/62660718/how-to-create-a-bar-graph-using-plotly-express-from-a-dictionary\n","\n","fig = px.bar(histogram, x = 'weights', y = 'count')\n","fig.show() "]},{"cell_type":"markdown","metadata":{},"source":["## Understanding Problem and Node"]},{"cell_type":"markdown","metadata":{},"source":["# Graph for genome relationships\n","\n","Construct a Graph class for the genomic relationships.\n","\n","*Note: by design, the nodes are identified by Key, not actual genome portion. This is for readability.*"]},{"cell_type":"code","execution_count":78,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["0 : {302: -5, 224: -16}\n","1 : {511: -5, 344: -6, 228: -18}\n","2 : {223: -22}\n","3 : {548: -5, 355: -24}\n","4 : {436: -17}\n","5 : {102: -5, 156: -5, 185: -15}\n","6 : {22: -7, 485: -7, 433: -12}\n","7 : {231: -25}\n","8 : {256: -6, 590: -23}\n","9 : {43: -20}\n"]}],"source":["class Graph:\n","    \"\"\"\n","    Taken from AIME4e, search.py \n","    Modified to include __repr__. \n","\n","    A graph connects nodes (vertices) by edges (links). Each edge can also\n","    have a length associated with it. \n","    The constructor call is something like:\n","        g = Graph({'A': {'B': 1, 'C': 2})\n","    this makes a graph with 3 nodes, A, B, and C, with an edge of length 1 from\n","    A to B,  and an edge of length 2 from A to C. You can also do:\n","        g = Graph({'A': {'B': 1, 'C': 2}, directed=False)\n","    This makes an undirected graph, so inverse links are also added. The graph\n","    stays undirected; if you add more links with g.connect('B', 'C', 3), then\n","    inverse link is also added. You can use g.nodes() to get a list of nodes,\n","    g.get('A') to get a dict of links out of A, and g.get('A', 'B') to get the\n","    length of the link from A to B. 'Lengths' can actually be any object at\n","    all, and nodes can be any hashable object.\"\"\"\n","\n","    def __init__(self, graph_dict=None, directed=True):\n","        self.graph_dict = graph_dict or {}\n","        self.directed = directed\n","        if not directed:\n","            self.make_undirected()\n","\n","    def make_undirected(self):\n","        \"\"\"Make a digraph into an undirected graph by adding symmetric edges.\"\"\"\n","        for a in list(self.graph_dict.keys()):\n","            for (b, dist) in self.graph_dict[a].items():\n","                self.connect1(b, a, dist)\n","\n","    def connect(self, A, B, distance=1):\n","        \"\"\"Add a link from A and B of given distance, and also add the inverse\n","        link if the graph is undirected.\"\"\"\n","        self.connect1(A, B, distance)\n","        if not self.directed:\n","            self.connect1(B, A, distance)\n","\n","    def connect1(self, A, B, distance):\n","        \"\"\"Add a link from A to B of given distance, in one direction only.\"\"\"\n","        self.graph_dict.setdefault(A, {})[B] = distance #🔴 What is setdefault? \n","\n","    def get(self, a, b=None):\n","        \"\"\"Return a link distance or a dict of {node: distance} entries.\n","        .get(a,b) returns the distance or None;\n","        .get(a) returns a dict of {node: distance} entries, possibly {}.\"\"\"\n","        links = self.graph_dict.setdefault(a, {})\n","        if b is None:\n","            return links\n","        else:\n","            return links.get(b)\n","\n","    def nodes(self):\n","        \"\"\"Return a list of nodes in the graph.\"\"\"\n","        s1 = set([k for k in self.graph_dict.keys()])\n","        s2 = set([k2 for v in self.graph_dict.values() for k2, v2 in v.items()])\n","        nodes = s1.union(s2)\n","        return list(nodes)\n","    \n","    def __repr__ (self): \n","        print (self.graph_dict)\n","\n","\n","genome_graph = Graph()\n","\n","# Note:  nodes can be added simply via connect, the connect function will take care of new vs existing nodes\n","for line in graph_adj_list:  #🔴 Can be cleaned up to replace graph_adj_list\n","    genome_graph.connect(line[0], line[1], line[2])\n","\n","# Verification \n","for node in genome_graph.nodes(): \n","    if node < 10: \n","        print (f'{node} : {genome_graph.get(node)}')\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["# Problem Class to solve genomic sequencing \n","\n","Similar to TSP"]},{"cell_type":"code","execution_count":79,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["TSP_problem([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 450, 451, 452, 453, 454, 455, 456, 457, 458, 459, 460, 461, 462, 463, 464, 465, 466, 467, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477, 478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 497, 498, 499, 500, 501, 502, 503, 504, 505, 506, 507, 508, 509, 510, 511, 512, 513, 514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 526, 527, 528, 529, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539, 540, 541, 542, 543, 544, 545, 546, 547, 548, 549, 550, 551, 552, 553, 554, 555, 556, 557, 558, 559, 560, 561, 562, 563, 564, 565, 566, 567, 568, 569, 570, 571, 572, 573, 574, 575, 576, 577, 578, 579, 580, 581, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598], None)\n"]}],"source":["# Implement subclass of Problem to solve the genome sequencing. \n","# First attempt: using Hill Climbing \n","import random\n","\n","class TSP_problem(Problem):\n","\n","    '''\n","    Source: obsolete_search4e.ipynb #🔴issue with animation \n","    subclass of Problem to define various functions \n","    '''\n","\n","    def two_opt(self, state):\n","        '''\n","        Neighbour generating function for Traveling Salesman Problem\n","        '''\n","        state2 = state[:]\n","        l = random.randint(0, len(state2) - 1)\n","        r = random.randint(0, len(state2) - 1)\n","        if l > r:\n","            l, r = r,l\n","        state2[l : r + 1] = reversed(state2[l : r + 1])\n","        return state2\n","\n","    def actions(self, state):\n","        '''\n","        action that can be excuted in given state\n","        '''\n","        return [self.two_opt]\n","    \n","    def result(self, state, action):\n","        '''\n","        result after applying the given action on the given state\n","        '''\n","        return action(state)\n","\n","    def path_cost(self, c, state1, action, state2):\n","        '''\n","        total distance for the Traveling Salesman to be covered if in state2\n","        '''\n","        cost = 0\n","        for i in range(len(state2) - 1):\n","            cost += distances[state2[i]][state2[i + 1]]\n","        cost += distances[state2[0]][state2[-1]]\n","        return cost\n"," \n","    def value(self, state):\n","        '''\n","        value of path cost given negative for the given state\n","        '''\n","        return -1 * self.path_cost(None, None, None, state)\n","\n","\n","initial = genome_graph.nodes() #this should be a path through the genomic reads. Expressed as a list. e.g. [1,3,5,12,32,...,23]. \n","tsp_problem = TSP_problem(initial)\n","\n","print (tsp)"]},{"cell_type":"code","execution_count":85,"metadata":{},"outputs":[{"ename":"AttributeError","evalue":"'TSP_problem' object has no attribute 'state'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)","\u001b[1;32m/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment 1 Q2-1.ipynb Cell 15\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=55'>56</a>\u001b[0m         \u001b[39mif\u001b[39;00m delta_e \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m \u001b[39mor\u001b[39;00m probability(\u001b[39m2.718281\u001b[39m\u001b[39m*\u001b[39m\u001b[39m*\u001b[39m(delta_e \u001b[39m/\u001b[39m T)):\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=56'>57</a>\u001b[0m             current \u001b[39m=\u001b[39m next_choice\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=59'>60</a>\u001b[0m states \u001b[39m=\u001b[39m simulated_annealing_full(tsp_problem)\n","\u001b[1;32m/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment 1 Q2-1.ipynb Cell 15\u001b[0m in \u001b[0;36msimulated_annealing_full\u001b[0;34m(problem, schedule)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=48'>49</a>\u001b[0m \u001b[39mif\u001b[39;00m T \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=49'>50</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m states\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=50'>51</a>\u001b[0m neighbors \u001b[39m=\u001b[39m current\u001b[39m.\u001b[39;49mexpand(problem)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=51'>52</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m neighbors:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=52'>53</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m current\u001b[39m.\u001b[39mstate\n","\u001b[1;32m/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment 1 Q2-1.ipynb Cell 15\u001b[0m in \u001b[0;36mNode.expand\u001b[0;34m(problem, node)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mexpand\u001b[39m(problem, node):\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mExpand a node, generating the children nodes.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m     s \u001b[39m=\u001b[39m node\u001b[39m.\u001b[39;49mstate\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m     \u001b[39mfor\u001b[39;00m action \u001b[39min\u001b[39;00m problem\u001b[39m.\u001b[39mactions(s):\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lenn/Documents/GitHub/IT5005-AI_AG3/Assignment%201%20Q2-1.ipynb#X23sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m         s1 \u001b[39m=\u001b[39m problem\u001b[39m.\u001b[39mresult(s, action)\n","\u001b[0;31mAttributeError\u001b[0m: 'TSP_problem' object has no attribute 'state'"]}],"source":["# Simulated Annealing stuff \n","\n","def hill_climbing(problem):\n","    \"\"\"\n","    [Figure 4.2]\n","    From the initial node, keep choosing the neighbor with highest value,\n","    stopping when no neighbor is better.\n","    \"\"\"\n","    current = Node(problem.initial)\n","    while True:\n","        neighbors = current.expand(problem)\n","        if not neighbors:\n","            break\n","        neighbor = argmax_random_tie(neighbors, key=lambda node: problem.value(node.state))\n","        if problem.value(neighbor.state) <= problem.value(current.state):\n","            break\n","        current = neighbor\n","    return current.state\n","\n","\n","def exp_schedule(k=20, lam=0.005, limit=100):\n","    \"\"\"One possible schedule function for simulated annealing\"\"\"\n","    return lambda t: (k * 2.718281**(-lam * t) if t < limit else 0)\n","\n","\n","def simulated_annealing(problem, schedule=exp_schedule()):\n","    \"\"\"[Figure 4.5] CAUTION: This differs from the pseudocode as it\n","    returns a state instead of a Node.\"\"\"\n","    current = Node(problem.initial)\n","    for t in range(sys.maxsize):\n","        T = schedule(t)\n","        if T == 0:\n","            return current.state\n","        neighbors = current.expand(problem)\n","        if not neighbors:\n","            return current.state\n","        next_choice = random.choice(neighbors)\n","        delta_e = problem.value(next_choice.state) - problem.value(current.state)\n","        if delta_e > 0 or probability(2.718281**(delta_e / T)):\n","            current = next_choice\n","def simulated_annealing_full(problem, schedule=exp_schedule()):\n","    \"\"\" This version returns all the states encountered in reaching \n","    the goal state.\"\"\"\n","    states = []\n","    current = Node(problem.initial)\n","    for t in range(sys.maxsize):\n","        states.append(current.state)\n","        T = schedule(t)\n","        if T == 0:\n","            return states\n","        neighbors = current.expand(problem)\n","        if not neighbors:\n","            return current.state\n","        next_choice = random.choice(neighbors)\n","        delta_e = problem.value(next_choice.state) - problem.value(current.state)\n","        if delta_e > 0 or probability(2.718281**(delta_e / T)):\n","            current = next_choice\n","\n","\n","states = simulated_annealing_full(tsp_problem)"]}],"metadata":{"colab":{"collapsed_sections":[],"provenance":[]},"kernelspec":{"display_name":"Python 3.10.1 64-bit","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.1"},"vscode":{"interpreter":{"hash":"aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"}}},"nbformat":4,"nbformat_minor":0}
